### ISSUE TITLE
Morphological terms in taxon pages not linked to terms in the Anatomical Glossary

### ISSUE THREAD

**Issue by REHarbach** [NaturalHistoryMuseum/scratchpads2#6550]
Morphological terms in taxon pages are not linked to the terms in the Anatomical Glossary. Also, the identifiers [no-lexicon] and [/no-lexicon] are showing in the text where morphological terms are not supposed to be linked to terms in the glossary. This issue seems to be linked to the previous issue I submitted regarding the problem with illustrations not being displayed in the Anatomical Glossary. These problems apparently stem from the site being moved to https: and myspecies. If possible please have a look at the site before the transition to see the links between text and glossary. These links were automatically created by the software in the past.

**Comment by Archilegt** [comment]
This problem was reported by me two years ago, on 13 May 2020:  Autotagging of taxon descriptions with terms from lexicon #6220 
Morphological term autotagging is one of the most important features of Scratchpads.

**Comment by therobyouknow** [comment]
I've linked that other issue based on your above comment.  They may indeed be linked. My ongoing investigation in #6549 will eventually confirm.

> This issue seems to be linked to the previous issue I submitted regarding the problem with illustrations not being displayed in the Anatomical Glossary. 

https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6549

**Comment by therobyouknow** [comment]
Can you show me which pages need to be linked please?

![image](https://user-images.githubusercontent.com/1080646/181739352-2bca55a2-31c2-4f41-9db6-bdddbff64ab5.png)

**Comment by REHarbach** [comment]
I think I have not explained the problem properly. It is not a matter of linking pages. Before myspecies, all of the morphological terms in all of the taxon pages were automatically liked to the definitions of the terms in the Anatomical Glossary. In the case of terms that should not be linked, e.g. Root as a surname as opposed to "root" in the list of Adult terminology, the term was enclosed in “[no-lexicon] … [/no-lexicon]”. It was my understanding that the ScratchPad software was configured to automatically make the links. Each time I added a new taxon page, the anatomical terms were automatically linked, indicated by highlight in blue. I then was able to identify terms that should not have been linked and I was able to block the link by enclosing the term within “[no-lexicon] … [/no-lexicon]”.

I thought the linking feature was common to all sites, but I now wonder if it was unique to my site. The original programmer for ScatchPads, Simon Rycroft, designed my sight, which has features that are not standard for ScratchPads.

I hope this will help to re-establish the links between anatomical terms in the taxon pages and those in the Anatomical Glossary.

Thank you for further consideration of this problem.

Ralph


Ralph E. Harbach

Scientific Associate

Natural History Museum

London SW7 5BD, UK

Tel. +44 (0)20 7942 5768

Fax. +44 (0)20 7942 2559



http://www.nhm.ac.uk/our-science/departments-and-staff/staff-directory/ralph-harbach.html<http://www.nhm.ac.uk/research-curation/stall-directory/entomology/r-harbach/index.html>

https://mosquito-taxonomic-inventory.myspecies.info/
________________________________

**Comment by therobyouknow** [comment]
Thank you @REHarbach 

I'm very grateful for the further information.  I certainly appreciate there's more to it than linking pages.

I can see that it is terms _within_ these pages that need to be linked, rather than just the overall pages.

https://mosquito-taxonomic-inventory.myspecies.info/simpletaxonomy/term/6427
https://mosquito-taxonomic-inventory.myspecies.info/node/11027

**Comment by Archilegt** [comment]
Hi, Rob. The autotagging feature is common to all Scratchpads.
Please read section "Semantic enrichment (tagging)" in:
Smith, V.S., Rycroft, S.D., Harman, K.T. et al. Scratchpads: a data-publishing framework to build, share and manage information on the diversity of life. BMC Bioinformatics 10, S6 (2009). https://doi.org/10.1186/1471-2105-10-S14-S6
The Taxonomy Autotagger module can be found at https://www.drupal.org/project/autotag

How it is meant to work: A user creates a glossary (= lexicon) which, except for the more advanced taxonomy for scientific names, is essentially a Drupal taxonomy. The user enters terms "head" and "leg" in a glossary, e.g., a morphology glossary. Then the user creates taxon pages via the Taxon Description content type, and includes text describing head and leg features. If configured correctly, upon clicking on the TD save button, the autotagging will be triggered and call existing glossaries. If there is a word match, the word in the taxon page will be highlighted in blue and enriched with a link to the corresponding glossary term.

Manual configuration is done through a separate interface. See  #6220 for partial description.

**Comment by Archilegt** [comment]
Tagging is partially working, what might be wrong is the configuration interface. It seems that some options have been lost.
Testing in Myriatrix: The taxon description for Chilopoda already existed. I edited the description, section "Morphology", with the following text: "Body composed of head and trunk." See https://myriatrix.myspecies.info/taxonomy/term/7151/descriptions
The terms "body", "head", and "trunk" already existed in the "Morphology" lexicon/glossary. See https://myriatrix.myspecies.info/nontaxonomy/term/38678
Results:
**1) Desired result not achieved.** The words in section "Morphology" (https://myriatrix.myspecies.info/taxonomy/term/7151/descriptions) are not highlighted and do not link to the "Morphology" glossary.
**2) Undesired result:** The words "body", "head", and "trunk" now appeared in the former blue banner Chilopoda, which now reads "Chilopoda, Body, Head, Trunk" in capital letters. This is the natural result of an unnatural configuration. When configuring the Taxon Description content type via admin/structure/types/manage/spm -> Tag options, the "Field to populate *" is a required field but only has one option: "Taxonomic name". Together with option "Append", the settings cause the undesired result in 2.
Expected visual behavior: 
1) The words should be highlighted in blue and directly linked to terms in the relevant glossaries.
2) Blue banners should not be affected by the "Field to populate *" option. Either make it optional or add options that are invisible or can be configured for visibility.
Rob, please notify me when you are done inspecting what I wrote about Myriatrix. I will then change the settings or delete the text because I don't want the blue banner Chilopoda to be affected.

**Comment by Archilegt** [comment]
The AntKey Scratchpad used to be a flagship Scratchpad for illustrating how good taxon descriptions could be autotagged with morphological terms in a glossary. Until recently, that Scratchpad used to have a separate installation at AntKey.org, with the glossary terms correctly highlighted in taxon pages and the links to the glossary working. That site is now gone but I found https://antkey.myspecies.info which is mostly working but with the morphological glossary unreachable at http://antkey.org/lexicon/43 and the morphological terms not highlighted in taxon pages.

**Comment by therobyouknow** [comment]
Thank you very much @Archilegt this information is most appreciated!  It will certainly help in my analysis of how to fix.  I will read through thoroughly and then advise my approach based on my updated understanding. Thank you again.

**Comment by therobyouknow** [comment]
> Rob, please notify me when you are done inspecting what I wrote about Myriatrix. I will then change the settings or delete the text because I don't want the blue banner Chilopoda to be affected.

Thank you @Archilegt   I have captured screenshots of the pages you refer to for my own reference as I continue to look into this problem.

You can now change the settings back to how they were.

https://myriatrix.myspecies.info/node/1183/edit relates to https://myriatrix.myspecies.info/taxonomy/term/7151/descriptions

![image](https://user-images.githubusercontent.com/1080646/182176688-a097dbc3-a8f5-41d3-94ba-68f2a4b68d1f.png)

also:

https://myriatrix.myspecies.info/nontaxonomy/term/38678

![image](https://user-images.githubusercontent.com/1080646/182177637-8e45c1bd-5759-428f-9b3c-dc101f44dcda.png)


I've also re-read your comments and have an improved understanding of the issue as a result.

With that information I can now experiment with the settings on my personal setup and observe the results to help determine how to fix the problem.

**Comment by therobyouknow** [comment]
**ongoing analysis**

https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1199269706

> When configuring the Taxon Description content type via admin/structure/types/manage/spm -> Tag options, the "Field to populate *" is a required field but only has one option: "Taxonomic name".

Yes, observing that I also see this:

https://myriatrix.myspecies.info/admin/structure/types/manage/spm/auto_label


![image](https://user-images.githubusercontent.com/1080646/182196825-6a0e1782-b99f-4051-9adf-7e97d581bb49.png)


`[node:field_taxonomic_name]`

**Comment by therobyouknow** [comment]
**ongoing analysis**

Looking at tokens to see if we can select certain ones to help with solution.

For a full list of Tokens, go here:

https://myriatrix.myspecies.info/admin/help/token

then click on "Browse available tokens."

and you get:

![image](https://user-images.githubusercontent.com/1080646/182198954-05c34bac-b26d-4444-9c15-19dd1d2455ce.png)

Or direct URL:

https://myriatrix.myspecies.info/token/tree?options=%7B%22token_types%22%3A%22all%22%2C%22click_insert%22%3Afalse%2C%22show_restricted%22%3Atrue%2C%22recursion_limit%22%3A2%7D&token=aF_IkyMHOK1EKLWVkPNBJAnp8ayrHL4WcEjaA0pCb40


I will experiment with this but I think there is potential for code changes required, including for the autokey module.

**Comment by therobyouknow** [comment]
**ongoing analysis**

Antkey has the same configuration as [Myam](https://myriatrix.myspecies.info/admin/structure/types/manage/spm/auto_label)

See below:

https://antkey.myspecies.info/admin/structure/types/manage/spm/auto_label

![image](https://user-images.githubusercontent.com/1080646/182373220-336a45d5-dac3-4e29-9665-228bfeac4bc3.png)

Being the same and that @Archilegt you've said that Antkey was a good source of examples for tagging, I would think there is potentially a bug in the software.

Some questions for you @Archilegt please:

Do you know when it stopped working?

Can you show me existing pages on https://myriatrix.myspecies.info where the tagging has worked?

**Comment by therobyouknow** [comment]
**ongoing analysis**

Apologies @Archilegt  by "tag options", I can now see you were referring to the vertical tab on the left on this page:

https://myriatrix.myspecies.info/admin/structure/types/manage/spm


![image](https://user-images.githubusercontent.com/1080646/182387231-f06008cd-8a26-4894-bda5-f69081f69226.png)

> When configuring the Taxon Description content type via admin/structure/types/manage/spm -> Tag options, the "Field to populate *" is a required field but only has one option: "Taxonomic name".

I had though that the "field to populate" - "taxonomic name" corresponded to the token set here: https://myriatrix.myspecies.info/admin/structure/types/manage/spm/auto_label - which is `[node:field_taxonomic_name]`  so I temporarily changed it on https://myriatrix.myspecies.info/admin/structure/types/manage/spm/auto_label  to [node:field_morphology] to see if that field would then be available as a drop down below "field to populate", but not the case. My experimentation of the configuration there.

My previous 2 questions still apply, so let me know your replies to them.

**dev notes**

Meanwhile, I think I'm going to need to inspect and debug the code for https://myriatrix.myspecies.info/admin/structure/types/manage/spm/auto_label and its dependency https://www.drupal.org/project/tag modules 

These modules are open sourced in the Drupal community contrib ecosystem, however they actually appear to exist in the Scratchpads codebase not as contrib modules but as custom modules:

- `autotag` contrib module appears to be `auto_label` custom module in Scratchpads codebase
- `tag` module appears to be `tagtag` module in Scratchpads codebase

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

I set xdebug some breakpoints in the above custom modules on my personal copy of the site, but break points not hit. So will try to choose some other ones, then I can inspect the variables and data to help my analysis as to why this issue is happening.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

Succinct summary of how it works in previous comment on this ticket further up:

https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1199217147

Next steps:

- attempt to recreate this on a local copy of site
- debug

Once fixed, https://scratchpads.readthedocs.io/en/latest/index.html may need updating to include this feature, as I did not come across it in my use of this training.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

Followed the steps in https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1199217147

on personal local demo test training site:

created a biological taxonomy called fruit with terms orange, lime, lemon

created a Taxon Description node, expe

![image](https://user-images.githubusercontent.com/1080646/186695262-5ac152af-c95d-45a7-bf01-1ec4ed4c902b.png)

Saved.

![image](https://user-images.githubusercontent.com/1080646/186695753-6c9f600b-5be2-4e03-b48d-d247d35c13de.png)

noting message:

`Biological terms must be added using the classification editor.`

is that significant.

now to examine config and use xdebug and see also if there's info in this ticket and elsewhere about this used to work.

**Comment by therobyouknow** [comment]
Good morning. Continuing to work on this issue. Will now do the 'next steps' mentioned above.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

I've looked into this further to try to solve the issue.

I'll summarise what I understand the requirement is that is needed to be restored and then go on to comment on the situation with the Scratchpads software.

In summary, this issue is about the restoration of a facility to link words entered into a content field to taxonomy terms that have those names.

e.g.  
1- You have a taxonomy `fruit` with terms `apple`, `orange`, `pear`. 
2 - You create content of type `Taxon Description` and during creation of the content, In one of the `Long text` type fields (i.e. `General Description` field (`field_general_description`) or `Morphology` field (`field_morphology`), you write a sentence which includes one or more of the above terms, e.g. "I like fruit salad, especially when it has pear and orange". 
3- Click save of your `Taxon Description` content. 
4 - Then, on viewing the content, you look for your "I like fruit salad, especially when it has pear and orange" sentence and observe that pear and orange each link to the page for their respective taxonomy terms.

Now, looking at the Scratchpads software.  It looks like this facility has only been code developed `Taxonomic Name` / `field_taxonomic_name` in `Taxon Description`.  

This `field_taxonomic_name` field seems to behave like the title of the content which doesn't make sense to me as a field that would be parsed to embed links to terms in. Moreover the field type of this seemingly title field is a taxonomy term, rather than a text type field that would be parsed. So I am a bit puzzled as to what is going on here.

I also can imagine how this facility could have ever worked if this is the only field apparently supported in the code.

I have come to the above conclusion by analysing `_tagtag_fields_list` in `tagtag.module` and `scratchpads_taxonomic_name_field_tag_field_types_alter` in `scratchpads_taxonomic_name_field` module.

I could try to add code to support for `Long text` type fields such as `field_general_description` and `field_morphology`, which I will do.  However, I note that there is an entire Drupal custom module dedicated to the `field_taxonomic_name`:  `scratchpads_taxonomic_name_field`  which concerns me that there is more to it than it seems in attempting to support these 2 additional fields.

I have some specific questions for the end users, which I will ask separately in a subsequent smaller comment below, so that they don't have to read all of the above technical analysis before asking my question.

**Comment by therobyouknow** [comment]
@REHarbach @Archilegt 

I have some questions for you:

1. Which fields do you want this facility to be working for and in which content type?  Do you want it working for the General Description field and the Morphology field of Taxon Description?
2. Any other fields you want it working for?
3. Did this ever work properly for you?
4. Do you have some content where it has worked now?
5. I note that https://mosquito-taxonomic-inventory.myspecies.info/ doesn't have any content of type Taxon Description. Did you want this facility working for another content type?

**Comment by therobyouknow** [comment]
Just adding entries for the 2 fields didn't work I'm afraid. 

I tried adding these 2 lines:...

```
  $populate_fields['field_morphology'] = "Morphology";
  $populate_fields['field_general_description'] = "General Description";
```

...after this existing line:...

`  $populate_fields = _tagtag_fields_list($form['#node_type']->type, 'populate');`

in file `sites/all/modules/custom/tagtag/tagtag.module`

"Back to the drawing board" - I need to think further about how this works.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

2 things appear to be broken in the code which could be the root cause as to why the tagging is not working:

-  incorrect api file name - the api hook implementation file is named `tag.api.php` but the module name is `tagtag`. I suggest that the api hook implementation be renamed to `tagtag.api.php` such that the name reflects the same name as its module, just like in other modules I've studied, e.g.  views.  I think that Drupal core will then recognise the API file.
- missing function - `tag_on_save_tag_form_alter` is referred to in `sites/all/modules/custom/tagtag/modules/tag_on_save/tag_on_save.module`  but the function is not in the code base at all, from my searching.  Therefore, I have some doubt that the `tag_on_save_tag_update` could ever get called  - that function launches another function `tagtag_add_tags` which tags in the content as per this ticket requirement. xdebug breakpoints show this never gets called on saving Taxon Description content as would be required.

Next steps: further xdebug debugging to confirm and testing.  Then to formulate a possible solution and initially test.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

Following on from my analysis noted in my previous comment, mentioning missing and incorrectly name code, on the basis of this, I can't see how this functionality ever worked.

Therefore the outline of the solution is to add new code, that makes use of the Drupal core hook node save/update hook functions, instead of those currently. Those current seem to attempt the same thing - to launch the tagging code on a node save. This could be done with a standard hook save or update. I'll look into implementing this now, should be a small amount of code.

**Comment by Archilegt** [comment]
Recap on previous comment https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1199269706

> Tagging is partially working, what might be wrong is the configuration interface. It seems that some options have been lost. Testing in Myriatrix: The taxon description for Chilopoda already existed. I edited the description, section "Morphology", with the following text: "Body composed of head and trunk." See https://myriatrix.myspecies.info/taxonomy/term/7151/descriptions The terms "body", "head", and "trunk" already existed in the "Morphology" lexicon/glossary. See https://myriatrix.myspecies.info/nontaxonomy/term/38678 Results: **1) Desired result not achieved.** The words in section "Morphology" (https://myriatrix.myspecies.info/taxonomy/term/7151/descriptions) are not highlighted and do not link to the "Morphology" glossary. 

I now think that I may have missing another essential tagging feature. It seems that the Lexicon module may have built-in tagging features, but they seem not to be working either. Needs investigation.
What to investigate: In any site, try the path admin/config/system/lexicon
In the case of Myriatrix: https://myriatrix.myspecies.info/admin/config/system/lexicon
When prompted, see Lexicon Settings -> General settings -> **checkbox** `Show/mark lexicon terms even if there is no description`. See also the explanatory text that appear when hovering over that option.

![image](https://user-images.githubusercontent.com/60008120/192145997-660ce886-61bc-466f-b3ef-a931e9a24027.png)

The checkbox was previously unchecked in Myriatrix. I checked it, saved, and then tried to edit and save the Pauropoda taxon description (https://myriatrix.myspecies.info/taxonomy/term/6570/descriptions), to see if the tagging would be forced upon save. Nothing happened, but I am inclined to believe that the checkbox and its code may be part of the answer we need, e.g., using a given lexicon to automatically tag (= mark?) words throughout the website.

See also Lexicon Settings -> Term marking settings -> **checkbox** `Mark terms in content`. The drop-down menu Match type could also be tried, especially the option `Any substring`.

![image](https://user-images.githubusercontent.com/60008120/192146586-4392c6da-315a-41fb-bb05-b13316200894.png)

I hope that this brings us closer to the answer!

**Comment by therobyouknow** [comment]
- to check function not called - 1h
- remember where to look, assuming its in the tagtag api file, inspect the hook invoke, debug to see if it is seeing that function that's not being called - yes or no - it is or isn't being called - 2h
- open ended - call stacks. check include file. check .info file includes that function with php in it

3h

**Comment by Archilegt** [comment]
Additional study material as in https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6595#issuecomment-1265287480

> About the automatic taxon recognition in keywords: This is a feature that has been stable and it is working correctly as far as I know. It is doing what we are not managing other vocabularies to do: It brute-forces dictionary-based recognition of scientific names in the "Keywords" field and then takes those recognized terms and adds them as comma-separated values into the field "Taxonomic name". If we want other dictionaries and recognition routines to work, we need to provide output fields where the recognized terms can be added as cs values. I am just mentioning this here because the "Keyword" to "Taxonomic name" pair across Scratchpads represents the successful routine that we want to recreate.

**Comment by therobyouknow** [comment]
Dev setup issues will impact my ability to work on this today, by up to a few hours. Hopefully they can be resolved asap by me.

![image](https://user-images.githubusercontent.com/1080646/193576030-c5c9a242-5afc-44e6-84fe-741a0fddec0a.png)

**Comment by therobyouknow** [comment]
switching to another machine for now to unblock myself, out of my 5 computers I have with me.

**Comment by therobyouknow** [comment]
error on first machine was: 

![image](https://user-images.githubusercontent.com/1080646/193576721-53b83c38-c566-4b3d-bce1-5f8bbdf6f5b5.png)

will have a look at this issue later.

These issues aren't to do with this issue but rather my setup.  I'm logging them here for the record so I can give a reason as to why my response timescale is not as quick as I'd like.

**Comment by therobyouknow** [comment]
> remember where to look, assuming its in the tagtag api file, inspect the hook invoke, debug to see if it is seeing that function that's not being called - yes or no - it is or isn't being called - 2h

Analysing `tag_form_node_type_form_alter` in `sites/all/modules/contrib/tag/tag.module`

**Comment by therobyouknow** [comment]
I looked into `tag_form_node_type_form_alter` in `sites/all/modules/contrib/tag/tag.module` which has a number of `module_invoke_all` function calls, but breakpoint for first line of this function not being hit either, it would appear.

By the way I'm doing a node save where I expect this hook to be invoked.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

now, I'm getting somewhere. The other function I mentioned earlier is used in config, not a node save. This one below is one of those I think I should be looking at and it does indeed xdebug break on node save:

![image](https://user-images.githubusercontent.com/1080646/193602702-9dbaf8a1-cc86-4bf2-8c81-80e2f888ec1f.png)

`tag_form_alter` in `sites/all/modules/contrib/tag/modules/tag_on_save/tag_on_save.module`

Will continue to investigate.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

2 questions I have for myself for my analysis:

1 - this line looks incorrect:
  `if(isset($form['#node_edit_form']) && $form['#node_edit_form']){`

The `$form` array never contains that index id. according to what I observered with the xdebug breakpoints. The code inside the if never runs as a result. So we need to replace the index.

2 - what do we replace the index with. The debugger shows that there is a `spm` content type in one of the form arrays that one of the calls to this function is made. We need something more generic I'd say which covers all content types where we want to have the word to tag insertion.

To further re-affirm that this function needs work to fix, inside the if is: `$search_services = module_invoke_all('tag_ui_info');` which should call a hook in `tag_on_save` sub module that will then call the `tag_on_save_tag_update` word to tag code.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**


![image](https://user-images.githubusercontent.com/1080646/193613164-f595446a-3e67-4f60-a00a-c59833fc9845.png)

**Comment by therobyouknow** [comment]
![image](https://user-images.githubusercontent.com/1080646/193613434-a075ec74-715d-4a72-9a16-ed97c60af08a.png)

**Comment by therobyouknow** [comment]
![image](https://user-images.githubusercontent.com/1080646/193613689-32e34b0e-79fc-460a-9a8a-e2a61f005a99.png)

**Comment by therobyouknow** [comment]
> Additional study material as in https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6595#issuecomment-1265287480
> 
> > About the automatic taxon recognition in keywords: This is a feature that has been stable and it is working correctly as far as I know. It is doing what we are not managing other vocabularies to do: It brute-forces dictionary-based recognition of scientific names in the "Keywords" field and then takes those recognized terms and adds them as comma-separated values into the field "Taxonomic name". If we want other dictionaries and recognition routines to work, we need to provide output fields where the recognized terms can be added as cs values. am just mentioning this here because the "Keyword" to "Taxonomic name" pair across Scratchpads represents the successful routine that we want to recreate.

Thank you @Archilegt for advising of this existing functionality. Here's a picture of what I think you have said:

![image](https://user-images.githubusercontent.com/1080646/193794054-90d4e1ef-3f5e-4369-8a9c-c3ed829591a6.png)

Is this correct? Please show me an example of this on your site or someone elses.


And here is a picture of what I am fixing for this issue 6550 here:

![image](https://user-images.githubusercontent.com/1080646/193794237-0ccfd10d-cf01-4211-ad57-633901b314e3.png)


Any comments let me know. Thank you!

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

**potential root cause as to why tagging words not working (this issue)**

in function `tag_submit_form` the index used for `$search_services` array is not correct, because it is an index that doesn't exist in the array.

![image](https://user-images.githubusercontent.com/1080646/193813738-a2bb3d42-1d24-439e-8bf0-0b0bf57bb9e1.png)

The `$search_services` array contains the tagging code, having had `module_invoke_all` which calls `tag_on_save_tag_ui_info`

![image](https://user-images.githubusercontent.com/1080646/193816159-a84dd38b-0ab8-4ca9-9dfa-48a64e250a5c.png)

Next steps:
- fix the index

**Comment by REHarbach** [comment]
I confess this is all gobbly-gook to me. I do not understand the two questions. As for replacement of "index", this brings to mind the word "lexicon" that was used before the site was moved to https and myspecies. What changed when the site was moved?


Ralph E. Harbach

Scientific Associate

Natural History Museum

London SW7 5BD, UK

Tel. +44 (0)20 7942 5768

Fax. +44 (0)20 7942 2559



http://www.nhm.ac.uk/our-science/departments-and-staff/staff-directory/ralph-harbach.html<http://www.nhm.ac.uk/research-curation/stall-directory/entomology/r-harbach/index.html>

https://mosquito-taxonomic-inventory.myspecies.info/
________________________________

**Comment by therobyouknow** [comment]
Don't worry Ralph!  Apologies as the notes I'm making are technical.  This is so it is the quickest for me to pick up from where I last left off and to generally show that the issue is being worked on for your reassurance and for my bosses! :) 

I always precede each technical comment entry with the heading title **ongoing analysis, dev notes** - when you see this please feel free to ignore the comment entry.  If I have a question for the user like yourself, I'll indicate in the heading title.

Thank you!

**Comment by therobyouknow** [comment]
potential fix on release sprint development branch `rob/release-sprint-2022-10-03-2022-10-10`. Now to test.

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

**debug testing outcome - more coding work needed**

`tag_on_save_tag_form_alter` is the callback function name being constructed in `tagtag_form_alter`, which doesn't exist in the code base.

But `tag_on_save_tag_update` does exist which calls the `tagtag_add_tags` function, but I haven't found anything that calls it yet, so it won't run.

**Comment by therobyouknow** [comment]
honestly

```
function tag_on_save_tag_update($form, &$form_state){
  // Simplest of all the UI submit functions.  This simply calls the 
```

` * Simple module that allows a user to select when entities/fields should be`

why does calling it `simple` help? what is someone else looking at that supposed to think?! "Oh yeah that's simple then, ah so that's alright then".

Pointless comments..

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

getting somewhere - this code below where the breakpoint is hit, is called on clicking save on a node.

The debugger shows the `tag_on_save`  `tag_on_save_tag_update` function is present in the array.

<img width="1038" alt="image" src="https://user-images.githubusercontent.com/1080646/194111938-87e6242e-3145-4209-8bcc-c30492651f8f.png">

**Comment by therobyouknow** [comment]
**ongoing analysis, dev notes**

The code that adds the submit handler we need calling, isn't being called.  My thinking is that there is an incorrect assumption being made about the construction of names to index into these arrays.

I have to say, the coding style and concepts are rather pompous (UI services, search services etc).  Could do with being more down to earth. The comments are awful. 

<img width="960" alt="image" src="https://user-images.githubusercontent.com/1080646/194122420-9d8372b4-89de-46b3-ab5f-434c52209d14.png">

**Comment by therobyouknow** [comment]
Re-doing https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1246836931 too see if we can get the fields to be populated.

Commit `commit a715f73c330f7f7e0f68f45a803c40d19050f933 (HEAD -> rob/release-sprint-2022-10-03-2022-10-10, origin/rob/release-sprint-2022-10-03-2022-10-10)` results in:


<img width="957" alt="image" src="https://user-images.githubusercontent.com/1080646/194338644-b3f652db-5579-4e8c-84a2-66064e8abfb9.png">

So I'll select General Description, save and then re-save my node to see if it is populated as per https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1266720598

**Comment by therobyouknow** [comment]
This hasn't worked. I can't spend any more time on this ticket.

The code seems to be more complicated than necessary to do: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1266720598

And the config form in section 'Tag options' is not as straightforward as it should be:

`/admin/structure/types/manage/spm`

<img width="229" alt="image" src="https://user-images.githubusercontent.com/1080646/194341675-335cccfb-4029-4125-9e62-1e10d65f2ad0.png">

The form shows 2 fields:
1 - fields to search
2 - taxonomies
3 - fields to populate

If we're  doing https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1266720598 Why then do we need "1 - fields to search" . We only need the fields we want to link words in text to the tag and to know what tags to match with the text.   And, no, we don't need to over engineer with UX just to make the form clearer.


So I'm pausing the work on this ticket for now because
a) the form settings are not clear and appear to have additional information and don't explain why
b) the code is overly complex and reasonable attempts at repair have not worked
c) this has taken a large amount of time out of the sprint so I need to put time into other tickets

**Comment by therobyouknow** [comment]
**ongoing analysis, usage process understanding and dev notes**

Updated understanding following study of custom module tagtag contrib module counterpart https://www.drupal.org/project/autotag:

1. we specify the tags we want to search for in content fields
2. we specify which content fields we want to do that search within
3. we spefify the output as to where the tagged words in those fields, i.e the field to output to

Above understanding based on:

> An additional option is provided on a content type configuration page

 `admin/structure/types/manage/page`  `admin/structure/types/manage/spm`

> specify which text fields should be searched for terms in specified vocabularies, and which field should then be populated with the results

![image](https://user-images.githubusercontent.com/1080646/195033302-aa88f436-8425-4f3f-b1eb-9f88bc0871b2.png)

This updated understanding has let me revisit the code to continue to troubleshoot the issue rather than consider a new module.  Analysis and troubleshooting ongoing.

**Comment by REHarbach** [comment]
1.   As far as I can recall, the facility previously worked for all content type, except content under literature, valid species and What's new on the main menu.
  2.  I don't know what you mean by "fields". It should work for content in all content except as noted above.
  3.  Yes, this worked properly before being switched to myspecies and https. In cases where terms were not to be linked, e.g. the surname "Root". the word was enclosed in [no-lexicon] [/no-no-lexicon]. Those commands are now showing in places in the content.
  4.  As far as I know, it is not working anywhere in the site. I have not seen any words highlighted in blue to indicate a link, but I have seen many of the commands to prevent a link.
  5.  As I mentioned some time ago, my site was one of the very first to be developed, and for that reason it deviates from the standard format, i.e. the taxon pages are uniquely formatted. If I don't currently have "Taxon Description", then it is not needd.



Ralph E. Harbach

Scientific Associate

Natural History Museum

London SW7 5BD, UK

Tel. +44 (0)20 7942 5768

Fax. +44 (0)20 7942 2559



http://www.nhm.ac.uk/our-science/departments-and-staff/staff-directory/ralph-harbach.html<http://www.nhm.ac.uk/research-curation/stall-directory/entomology/r-harbach/index.html>

https://mosquito-taxonomic-inventory.myspecies.info/
________________________________

**Comment by Archilegt** [comment]
Hi, @therobyouknow! Sorry that I did not reply before, but it would have taken me a lot of typing.
I confirm that your latest comment is the correct one.
Something that is general: The process in the background is a dictionary-based recognition, where Drupal lexicons act as dictionaries. And again, remember that in Scratchpads you have two categories of lexicons: 1) the common Drupal lexicons, and 2) the taxonomic name lexicons of Scratchpads, which are highly modified Drupal lexicons. When you test the tagging and the interfaces, you have to test for both because the taxonomic lexicons have specific processes and modules designed for them.

Updating your graphs in [this comment](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1266720598) above:

![Scratchpads tagging](https://user-images.githubusercontent.com/60008120/195105386-14ee89ed-1694-4701-8ffc-b1f6288bc2bd.jpg)

I hope that this is useful.

**Comment by therobyouknow** [comment]
Very useful thank you @Archilegt and question: is Taxonomic Name the only output field that we want tagged words to go into?

(referring to your diagram and example: https://myriatrix.myspecies.info/content/invasive-species-cylindrodesmus-hirsutus-pocock-diplopoda-polydesmida-haplodesmidae-spreads )

Or other fields possible to be selected as well?

**Comment by therobyouknow** [comment]
**ongoing analysis, response to another user** 

@REHarbach  thank you very much for your [comment](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1274336025)I will get back to you if I have further questions.

**Comment by therobyouknow** [comment]
**ongoing analysis, breaking down understanding of related functionality**

Thank you @Archilegt I will focus on the issue with the tagging as this issue raises: 

<img width="399" alt="image" src="https://user-images.githubusercontent.com/1080646/195592303-0f1a4540-53ca-475e-a725-52a805471b5c.png">

Noted about the 2nd lower half of your diagram (shown below), whereby you say words from the field text can be added as new taxonomy terms. Are there also any problems with adding new tags from the field text?

<img width="400" alt="image" src="https://user-images.githubusercontent.com/1080646/195592535-e711184e-4b99-4ee3-9235-b82b1a990852.png">

**Comment by therobyouknow** [comment]
**ongoing analysis, help notes for dev for assessing mechanism**

Viewing content:
https://myriatrix.myspecies.info/content/invasive-species-cylindrodesmus-hirsutus-pocock-diplopoda-polydesmida-haplodesmidae-spreads

Editing above content:
https://myriatrix.myspecies.info/node/8/edit

**Comment by therobyouknow** [comment]
<img width="933" alt="image" src="https://user-images.githubusercontent.com/1080646/196705178-41208dcf-029b-49fd-89be-ed4dca5081cf.png">

**Comment by therobyouknow** [comment]
**ongoing analysis**

**experiment - successful useful to know mechanism**

on save, words in keywords field are used to populate taxonomy name and I observed this.  we deleted a few terms from taxonomy name, saved but they came back because they came from the keywords.

So task for dev is to isolate code that uses keywords in this way to populate taxonomy name, which will then inform how to fix this issue 6550 


<img width="927" alt="image" src="https://user-images.githubusercontent.com/1080646/196706274-d861839d-52cf-4dfd-94f3-b5ecde14a9c0.png">

**Comment by therobyouknow** [comment]
<img width="934" alt="image" src="https://user-images.githubusercontent.com/1080646/196712034-24728151-be38-445c-8059-2584aea5d4c8.png">

**Comment by therobyouknow** [comment]
**ongoing analysis**

Recapping on the [original issue comment](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issue-1245019876)

@REHarbach [said](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issue-1245019876):

> Morphological terms in taxon pages

By "taxon pages" we mean Taxon Description content type.

>  are not linked to the terms in the Anatomical Glossary.

By "are not linked" we mean from any rich text field (i.e. that supports hyperlinking).

And by "Anatomical Glossary" we are talking about Drupal Taxomony.

And "terms" are Drupal taxonomy terms.

"Anatomical Glossary" is any Drupal Taxonomy that is biological.

> Also, the identifiers [no-lexicon] and [/no-lexicon] are showing in the text 

(identifiers = tag)

My question is where? Give me URL web address as to where `[no-lexicon]` and `[/no-lexicon]` are showing, please.

Answer - example URL web address is: https://mosquito-taxonomic-inventory.myspecies.info/node/11158

<img width="726" alt="image" src="https://user-images.githubusercontent.com/1080646/196722746-17b78618-e720-4ef5-a2d7-b80d3971c72c.png">


> where morphological terms are not supposed to be linked to terms in the glossary.

Which means he doesn't want `[no-lexicon]` and `[/no-lexicon]` showing.

Carlos suggests characterising this issue by experimenting with https://myriatrix.myspecies.info/nontaxonomy/term/38679

> This issue seems to be linked to the previous issue I submitted regarding the problem with illustrations not being displayed in the Anatomical Glossary. 

Thank you but I think the above ^ comment about another issue is unrelated.

> These problems apparently stem from the site being moved to https: and myspecies. If possible please have a look at the site before the transition to see the links between text and glossary. 

Again thank you but I think the above ^ comment about another issue is unrelated.

> These links were automatically created by the software in the past.

We think this is 2007-2016

Carlos advises, Dimitris Koureas at Naturalis, Leiden, Netherlands may be able to advice generally on concepts, to explain how things are meant to work.

**Comment by therobyouknow** [comment]
**ongoing analysis**

<img width="931" alt="image" src="https://user-images.githubusercontent.com/1080646/196734696-80491a06-f103-4842-afd9-cd83ea16f1f1.png">


<img width="932" alt="image" src="https://user-images.githubusercontent.com/1080646/196734780-f4007f28-be99-475f-a6a4-c49a673d2bd7.png">

greenline 
<img width="547" alt="image" src="https://user-images.githubusercontent.com/1080646/196734963-723004e2-0015-4286-9d4d-49c4f11c9a87.png">


illegal values are highlighted

<img width="932" alt="image" src="https://user-images.githubusercontent.com/1080646/196735411-89b89303-d40c-4cb8-909d-440671feef92.png">

deleted highlighted values, then able to save ok (orange line in Carlos's diagram)

<img width="938" alt="image" src="https://user-images.githubusercontent.com/1080646/196735596-b190849a-2896-4fbb-af29-a4189ba6a2ba.png">

**Comment by therobyouknow** [comment]
We believe the following specific functionality is working: that A Drupal vocabulary has terms added it from the Keywords field if they don't already exist in the vocabulary. 

<img width="600" alt="image" src="https://user-images.githubusercontent.com/1080646/196737157-7454cbc8-de0f-4213-aa9a-b850de632bf4.png">

**Comment by therobyouknow** [comment]
From talking with Carlos video chat today.

The primary issue is that fields that are long text / rich text _need to have_ words that exist as drupal taxonomy terms  made into hyper links to those tags.

![image](https://user-images.githubusercontent.com/1080646/196738229-b2960b4d-7423-4075-b687-2059c16d1cf3.png)

**Comment by therobyouknow** [comment]
Carlos suggested to look here: https://myriatrix.myspecies.info/admin/config/system/lexicon

<img width="940" alt="image" src="https://user-images.githubusercontent.com/1080646/196744175-799b7734-6913-452d-bd86-df5ff6bc0627.png">

- Term marking settings
-- You have to enable the Lexicon Filter in one or more text formats in the 

- Text format settings
-- Mark terms in content.

- HideTerm matching
-- MATCH TYPE
--- Word

The above suggests the (broken) functionality is somewhere else I haven't yet seen so maybe a new module might not be necessary.

To find the above form in the code, I will search the text strings in the code.

**Comment by therobyouknow** [comment]
note also

https://myriatrix.myspecies.info/admin/config/content/biblio/keywords

<img width="951" alt="image" src="https://user-images.githubusercontent.com/1080646/196746333-22b65e7b-f745-4976-bcb4-0afd01445956.png">

**Comment by therobyouknow** [comment]
Great call with Carlos. 3h video webchat. Thank you for your help.

**Comment by therobyouknow** [comment]
Vladimir Blagoderov, Vlad also a good contact.

**Comment by therobyouknow** [comment]
ZooKeys 50. 2010. Taxonomy shifts up a gear

**Comment by therobyouknow** [comment]
ZooKeys 150. 2011. e-Infrastructures for data publishing in biodiversity science

**Comment by Archilegt** [comment]
The wording of the related [Tag API module](https://www.drupal.org/project/tag) may help in understanding the dual nature of the Taxonomy Autotagger module.

**Comment by therobyouknow** [comment]
These [settings](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1284248618) are inside the `lexicon` module, `sites/all/modules/contrib/lexicon`.

**Comment by REHarbach** [comment]
When will you next be in the museum. I am usually in on Mondays. I will need your help to understand and deal with these elements.

Regards,

Ralph


Ralph E. Harbach

Scientific Associate

Natural History Museum

London SW7 5BD, UK

Tel. +44 (0)20 7942 5768

Fax. +44 (0)20 7942 2559



http://www.nhm.ac.uk/our-science/departments-and-staff/staff-directory/ralph-harbach.html<http://www.nhm.ac.uk/research-curation/stall-directory/entomology/r-harbach/index.html>

https://mosquito-taxonomic-inventory.myspecies.info/
________________________________

**Comment by therobyouknow** [comment]
Hi Ralph @REHarbach I will be aiming for in the next 2 weeks or so.  I've had some family to look after and so I've had to work remotely most of the time recently.  But I hopefully over the next 2 weeks or so I think I'll be in a better position to start being onsite more often. Things are pointing positively in that direction. I certainly want to do that!  In meantime message me. I look forward to speaking with you, thank you!

**Comment by therobyouknow** [comment]
**ongoing analysis,   location of problem is likely `lexicon` module**

In summary, this issue is about the following not working:  a body of text containing taxonomy terms each which should be made into links to the term.

The `lexicon` module does this functionality not `autotag`, not `tagtag` etc.  It took me a lot of perseverence to get to this conclusion. Thanks to @Archilegt for bringing this to my attention on our 3h video chat last week.

The `lexicon` module does need fixing to make this functionality work again.

Here are the config admin settings that I believe should enable words within a body of text to be recognised as tags and for them to be each converted to a link to the term:

In `admin/config/system/lexicon`:

<img width="714" alt="image" src="https://user-images.githubusercontent.com/1080646/197419817-38ddabdd-9374-47f6-9eeb-ec863c266d0f.png">


<img width="652" alt="image" src="https://user-images.githubusercontent.com/1080646/197419823-d9af6cb7-fb78-49b2-ad44-b22cd355cf82.png">


note particularly

<img width="610" alt="image" src="https://user-images.githubusercontent.com/1080646/197419845-edb370e1-822e-4cf6-904a-1045ca7d6825.png">

in Filtered HTML admin/config/content/formats/filtered_html


<img width="571" alt="image" src="https://user-images.githubusercontent.com/1080646/197419870-8f093388-55cd-4ce4-9ece-e027de0d8a4a.png">

see

<img width="299" alt="image" src="https://user-images.githubusercontent.com/1080646/197419862-30f17315-b882-470a-ba74-4b51587f2f15.png">


Back to Lexicon Settings admin/config/system/lexicon



<img width="883" alt="image" src="https://user-images.githubusercontent.com/1080646/197419905-02d310f6-08a8-482e-84cb-1eeea9201793.png">

I think these settings should work, but they don't. As in, words in my test site that are taxonomy terms, are not converted to links to terms.

So my next step is to debug the `lexicon` module with xdebug upon saving content with long text fields containing words which are taxonomy terms / tags and then be able to find out what's going wrong.

**Comment by therobyouknow** [comment]
**ongoing analysis, documentation deficient**

I would also add that there doesn't appear to be information about the `lexicon` module in these official training materials:

https://www.google.com/search?q=site%3Ascratchpads.readthedocs.io%2Fen%2Flatest%2F+lexicon&oq=site%3A&aqs=chrome.0.69i59j69i57j69i58j69i65l3.1622j0j7&sourceid=chrome&ie=UTF-8

<img width="725" alt="image" src="https://user-images.githubusercontent.com/1080646/197493210-fa97d159-c59c-4a22-ba3b-4fe1c261b965.png">

**Comment by therobyouknow** [comment]
Debugging this error message seen on saving of content.

<img width="621" alt="image" src="https://user-images.githubusercontent.com/1080646/197506183-eb7c9c34-c698-4cf4-bc9e-2955fed681e5.png">


`Notice: Undefined property: stdClass::$processing in _lexicon_get_terms() (line 1090 of /var/www/html/sites/all/modules/contrib/lexicon/lexicon.module).`

**Comment by therobyouknow** [comment]
Enabling `Internationalization`  and  `i18n_taxonomy` modules, because currently `_lexicon_get_terms` returns empty array of terms, which can explain why the functionality that this issue is about, isn't working. `_lexicon_get_terms` has logic for when `Internationalization`  and  `i18n_taxonomy` modules are enabled and I want to see how this logic works and what it can do to help solve the issue.

<img width="152" alt="image" src="https://user-images.githubusercontent.com/1080646/197526721-1412a12f-748d-4dfa-a342-5bc7ab4d7c58.png">

**Comment by therobyouknow** [comment]
**ongoing analysis, lexicon module, analysis of running (using xdebug)**

With `Internationalization` / `i18n` and `Taxonomy translation` / `i18n_taxonomy` modules, the following logic can be active, I'll see what it does:

<img width="417" alt="image" src="https://user-images.githubusercontent.com/1080646/197530536-09298adf-9c4d-4df6-8577-94bef3660845.png">

**Comment by therobyouknow** [comment]
ongoing analysis, lexicon module, analysis of running (using xdebug)

BOOM!

This is now working @Archilegt  on my test site.  I'll do some clean up and further checks.

<img width="472" alt="image" src="https://user-images.githubusercontent.com/1080646/197536619-0bcec57e-8fb6-446e-b7a9-b33522a2ad0e.png">

commented out as follows in _lexicon_get_terms in lexicon module

<img width="211" alt="image" src="https://user-images.githubusercontent.com/1080646/197536709-cd54850a-e465-40de-99b7-9f44eb26a522.png">

**Comment by Archilegt** [comment]
<iframe src='https://gfycat.com/ifr/FrayedImprobableElephant' frameborder='0' scrolling='no' allowfullscreen width='640' height='545'></iframe><p><a href="https://gfycat.com/discover/celebrate-gifs">from Celebrate GIFs</a> <a href="https://gfycat.com/frayedimprobableelephant-excitement-jonah-hill-celebrate-screaming">via Gfycat</a></p>

**Comment by therobyouknow** [comment]
**solution candidate found, items defined for this solution to be functional**

With solution [found](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1289035623) here is what it does and what's needed for it to be active and working.

**what the solution fixes**

The [solution](https://github.com/NaturalHistoryMuseum/scratchpads2/commit/5a256ea7b50f0e9dba18211edd99da3b1bbe22d7) re-enables existing functionality that, on saving content, looks in the rich text fields for each words that is a Drupal taxonomy term and links the word to the matching term's page:

![image](https://user-images.githubusercontent.com/1080646/197562305-854bbdf6-9026-4bfa-a068-f369f5cddec1.png)

A Drupal taxonomy = Lexicon = Vocabulary = Glossary (to re-iterate the legend at the foot of this [comment](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6550#issuecomment-1274696368))

A Drupal taxonomy term = Lexical item = Vocabulary word = Glossary term.

**what is needed to make the solution active and working**

- The [solution's code fix](https://github.com/NaturalHistoryMuseum/scratchpads2/commit/5a256ea7b50f0e9dba18211edd99da3b1bbe22d7)
- Releasing this code fix to the sites that use this facility - date to be confirmed
- Enabling these 2 contrib modules: `Internationalization` / `i18n` and `Taxonomy translation` / `i18n_taxonomy`.
- Configuring the `lexicon` module in Lexicon Settings at: `admin/config/system/lexicon` - here is what I configured for my test site where I observed the solution to work:

<img width="540" alt="image" src="https://user-images.githubusercontent.com/1080646/197568075-3b28991f-15ce-4966-8801-1d76b162d3e0.png">

- Configure one or more Text Formats to use the functionality - check/tick/enable box for  `Mark Lexicon terms`:

My example shows 'Filtered HTML` text format:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/1080646/197568196-e4938673-ce86-4ae3-9b58-088ae2de316b.png">

Use the text format in the content, e.g. `Filtered HTML`

<img width="666" alt="image" src="https://user-images.githubusercontent.com/1080646/197568763-19992222-c611-464a-84b7-b90e224c4c3d.png">
